### PR TITLE
Create projectComplete() dependency method in JvmComponentDepenencies

### PR DESCRIPTION
**Pull request by jvandort**
Adds a method which creates a dependency on the current project, except with an additional CompileView#JAVA_COMPLETE attribute.

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test qfl

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java [hunk 1]
```diff
@@ -125,4 +125,15 @@ public interface JvmComponentDependencies {
...
+    /**
+     * Create a dependency on the current project's complete view. During compile-time, this dependency will
+     * resolve the current project's implementation in addition to its API. During runtime, this dependency
+     * behaves as a usual project dependency.
+     *
+     * @return A new dependency
+     *
+     * @since 7.6
+     */
+    ModuleDependency projectComplete();
```
Comment (octylFractal) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java:138]: This is a `ProjectDependency`, right? I would prefer to be as specific as possible in the type.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmComponentDependencies.java [hunk 2]
```diff
@@ -83,6 +85,19 @@ public Dependency testFixtures(ModuleDependency moduleDependency) {
         return moduleDependency;
     }
 
+    @Inject
+    protected abstract Project getCurrentProject();
+
+    @Inject
+    protected abstract ObjectFactory getObjectFactory();
```
Comment (octylFractal) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmComponentDependencies.java:92]: These should be next to `getDependencyFactory()`, and `getDependencyFactory()` should be made `abstract` to match.

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -339,6 +339,25 @@ one of the addresses of the current machine's network interfaces.
 Similarly, the new Gradle property `org.gradle.debug.host` now enables [running the Gradle process with the debugger server](userguide/troubleshooting.html#sec:troubleshooting_build_logic)
 accepting connections via network on Java 9+.
 
+#### Introduced `projectComplete()` dependency for creating unit-like test suites
+
+The [JVM test suite](userguide/jvm_test_suite_plugin.html) `dependencies` block now has support for depending on the complete view of the current project during compile-time. Previously it was only possible to depend on the current project's API. This allows declaration of unit-like test suites which have access to dependencies of project internals which are not declared on the `api` or `compileOnlyApi` configurations.
```
Comment (octylFractal) [inline subprojects/docs/src/docs/release/notes.md:344]: > which have access to dependencies of project internals which are not declared on the `api` or `compileOnlyApi` configurations. 

This is hard to parse for me due to the double use of `which`. I can't think of a good improvement though, perhaps @nathan-contino has some ideas.

**Review by octylFractal — APPROVED** [review]
LGTM, I have one potential doc issue though.

File: subprojects/docs/src/docs/release/notes.md [hunk 4]
```diff
@@ -339,6 +339,25 @@ one of the addresses of the current machine's network interfaces.
 Similarly, the new Gradle property `org.gradle.debug.host` now enables [running the Gradle process with the debugger server](userguide/troubleshooting.html#sec:troubleshooting_build_logic)
 accepting connections via network on Java 9+.
 
+#### Introduced `projectComplete()` dependency for creating unit-like test suites
+
+The [JVM test suite](userguide/jvm_test_suite_plugin.html) `dependencies` block now has support for depending on the complete view of the current project during compile-time. Previously it was only possible to depend on the current project's API. This allows declaration of unit-like test suites which have access to dependencies of project internals which are not declared on the `api` or `compileOnlyApi` configurations. 
+
+For example, the following snippet uses the new `projectComplete()` API to define a new unit-like test suite:
```
Comment (tresat) [inline subprojects/docs/src/docs/release/notes.md:346]: I don't like the name `unit-like` here - what kind of tests are these?  I think you should briefly describe a case where you would want to access the internals and name the suite accordingly.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java [hunk 5]
```diff
@@ -125,4 +125,15 @@ public interface JvmComponentDependencies {
...
+    /**
+     * Create a dependency on the current project's complete view. During compile-time, this dependency will
+     * resolve the current project's implementation in addition to its API. During runtime, this dependency
+     * behaves as a usual project dependency.
+     *
+     * @return A new dependency
+     *
+     * @since 7.6
+     */
+    ProjectDependency projectComplete();
```
Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java:138]: I've been picking on names lately, apologies in advance.

Reading this and seeing the examples, I think this is name misleading, as it implies the usual `project()` dep is somehow incomplete.  I think you'll get newcomers finding it and thinking, "well, if I have the choice, I definitely want to depend on the _complete_ project, right?"  What happened to `projectInternal()` or `projectImplementation()`?

**Review by tresat — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/docs/release/notes.md (hunk 3 above)
Comment (nathan-contino) [inline subprojects/docs/src/docs/release/notes.md:344]:
```suggestion
#### Introduced `projectComplete()` dependency for test suites with access to project internals

The [JVM test suite](userguide/jvm_test_suite_plugin.html) `dependencies` block now
supports depending on the complete view of the current project during compile-time.
Previously it was only possible to depend on the current project's API. This allows
test suites with access to project internals that are not declared on
the `api` or `compileOnlyApi` configurations.
```
Suggestion: minor rewords, remove "unit-like" since it's an unfamiliar term that we're unlikely to use in the future.
Added explicit newlines for sanity's sake.

File: subprojects/docs/src/docs/release/notes.md (hunk 4 above)
Comment (nathan-contino) [reply subprojects/docs/src/docs/release/notes.md:346]:
```suggestion
For example, the following snippet uses the new `projectComplete()` API to define a test suite with access to project internals:
```
Suggestion: use the definition of "unit-like" instead of the term.

**Review by nathan-contino — APPROVED** [review]
Docs LGTM -- added some suggestions for clarity's sake.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java (hunk 5 above)
Comment (big-guy) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java:138]: +1 for projectImplementation 

other ideas:
- projectForUnitTesting()
- projectWithImplementation()
- clearboxProject()
- unitTestableProject()

**Review by big-guy — APPROVED** [review]
Please rebase this once the other PR gets merged.

I'm good with this except for the name of the new method, so please come to some agreement with @tresat on it.

I would also like to see an example put into `jvm_test_suite_plugin.adoc` for declaring a custom "unit test" test suite

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java (hunk 5 above)
Comment (jvandort) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java]: The naming for this one has been particularly challenging. At this point, we've mostly narrowed down the options to `projectInternalView()` and `projectInternals()`, with the former being the current state of the PR. Below are some of the other options considered along the way and the reasonings for dismissing them:

1. `projectComplete()`: Based off the attribute value used for the `CompileView` which differentiates between `apiElements` and `compileElements`. This could potentially confuse users who would may assume the "complete" view is better than the default one. 
2. `projectImplementation()`: Too easily confused with the `implementation` configuration. Users could assume they are getting dependencies only from the project's `implementation` configuration. This overloads the term "Implementation". 
3. `projectInternal()`: Too easily confused with `ProjectInternal`
4. `projectCompile()`: Can be confused as providing the project's compile classpath and only the compile classpath. This dependency is in reality no different than the `project()` dependency during runtime, so having "compile" in the name seems wrong. 

In response to the additional posted ideas:
- IMO the clearbox/opaquebox naming scheme seems forced. I'd prefer to steer away for now
- I would like to avoid linking this dependency to unit testing, since people will certainly figure out ways to use this dependency for suite types which aren't necessarily unit tests. 

I am still open to ideas, and test suites are still incubating. I am pretty satisfied with `projectInternalView()`, though `projectInternals()` is a close second which I would be content with. Thoughts?

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/ProjectInternalViewDependencyIntegrationTest.groovy [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+        when:
+        succeeds "customTest"
+
+        then:
+        result.assertTaskExecuted(":jar")
+        result.assertTaskExecuted(":customTest")
+    }
+
+    def writeBaseBuildFile() {
```
Comment (tresat) [inline subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/ProjectInternalViewDependencyIntegrationTest.groovy:78]: Split this into 2 methods with better names?  I'd have a `setup` method write the build file's `plugins`, `repos` and `testing` blocks, then only do the test specific top-level `dependencies` blocks in each test's `given`.  Then this can be renamed `writeSource()`.  It's moderately nitpicky, but I think it would be clearer - this method's name is a little misleading.

**Review by tresat — APPROVED** [review]

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmComponentDependencies.java [hunk 7]
```diff
@@ -82,6 +83,13 @@ public ProjectDependency project() {
         return getDependencyFactory().create(getCurrentProject());
     }
 
+    @Override
+    public ProjectDependency projectInternalView() {
+        return (ProjectDependency) project().attributes(attrs -> {
+            attrs.attribute(CompileView.VIEW_ATTRIBUTE, getObjectFactory().named(CompileView.class, CompileView.JAVA_COMPLETE));
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmComponentDependencies.java:89]: Yes I think we should. Not gonna do that in this PR, but I plan to put out another PR (maybe on master at this point, which should be fine given the attribute is incubating), to change `JAVA_COMPLETE` to `JAVA_INTERNAL`

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/ProjectInternalViewDependencyIntegrationTest.groovy (hunk 6 above)
Comment (jvandort) [reply subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/ProjectInternalViewDependencyIntegrationTest.groovy:78]: The problem with this is that each test uses a different plugin, so their `plugins` block must be different, and therefore the `setup` method can't write to the build-script.

File: subprojects/docs/src/docs/userguide/jvm/jvm_test_suite_plugin.adoc [hunk 8]
```diff
@@ -139,6 +139,7 @@ include::sample[dir="snippets/testing/test-suite-configure-suite-dependencies/ko
 <1> Configure the built-in `test` test suite.
 <2> Add the assertj library to the test's compile and runtime classpaths.  The `dependencies` block within a test suite is already scoped for that test suite.  Instead of having to know the global name of the configuration, test suites have a consistent name that you use in this block for declaring `implementation`, `compileOnly`, `runtimeOnly` and `annotationProcessor` dependencies.
 <3> Add the Auto Value annotation processor to the suite's annotation processor classpath so that it will be run when compiling the tests.
+<4> Add the `projectInternalView()` dependency to the integration test's compile and runtime classpaths. This dependency allows a test suite to access internal project dependencies which are declared on the project's `implementation` and `compileOnly` configurations, in addition to the `api` or `compileOnlyApi` configurations that the `project()` dependency provides.
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/docs/src/docs/userguide/jvm/jvm_test_suite_plugin.adoc:142]: Sounds good!

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by nathan-contino — APPROVED** [review]
Docs LGTM

**Comment by jvandort** [comment]
@bot-gradle test and merge

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java (hunk 5 above)
Comment (ljacomet) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmComponentDependencies.java]: I find myself in agreement with the dismissed names, Justin provides good reasons for not using them.

For the two candidates, my preference is on `projectInternalView()`.

One thing that could be done is actually name it `internalProjectView()` so that it is easy to find if you know about it and know you need it. But it also does not come up as a suggestion when you start typing `project`.

I also cannot find an alternative proposal that makes sense at this point.
